### ISSUE TITLE
Jobs are dropped from Sidekiq queues

### ISSUE THREAD

**Issue by egeek** [mperham/sidekiq#5614]
Ruby version: 3.1.2
Rails version: 7.0.4
Sidekiq: 6.5.7

We're experiencing the same issue, we batch queue a large number of jobs, some of them process, then after a while (maybe 30 minutes) the jobs suddenly start processing with next to no time, but it's not actually calling the code in the job. We have a log statement at the start of the perform function in the job but it's not getting called.

```
Nov 01 13:40:27Z <APP> app/<WORKER>.1 pid=4 tid=hmic class=<MODULE>::<JOB> jid=43bc4c583af46a1103d4a873 INFO: start
Nov 01 13:40:27Z <APP> app/<WORKER>.1 pid=4 tid=hmic class=<MODULE>::<JOB> jid=43bc4c583af46a1103d4a873 elapsed=0.003 INFO: done
```

This job should take around 10 seconds to execute.

Once this starts happening sidekiq processes all the remaining jobs in the queue (tens of thousands) in this way, without actually calling the perform function, until the queue is empty.

Sidekiq YML file for the worker in question.
```
:concurrency: 5
:max_retries: 1
:queues:
  - [ my_batch, 4 ]
```

sidekiq.rb
```
require "sidekiq"
require "sidekiq-scheduler"
require "sidekiq-status"
require "sidekiq/web"
require "sidekiq-scheduler/web"
require "sidekiq/throttled"
require "sidekiq/middleware/server/sidekiq_exception_handler_adapter"

Sidekiq.default_job_options = {
  backtrace: true,
  retry: true
}

Sidekiq.configure_client do |config|
  config.redis = ConnectionPool.new(timeout: 30, size: 100) {
    Redis.new(url: Rails.configuration.redisgreen[:url], driver: :hiredis)
  }

  config.client_middleware do |chain|
    chain.add SidekiqUniqueJobs::Middleware::Client
    chain.add Sidekiq::Status::ClientMiddleware, expiration: 30.minutes
  end
end

Sidekiq.configure_server do |config|
  config.redis = ConnectionPool.new(timeout: 30, size: 100) {
    Redis.new(url: Rails.configuration.redisgreen[:url], driver: :hiredis)
  }

  config.server_middleware do |chain|
    chain.add Sidekiq::Middleware::Server::SidekiqExceptionHandlerAdapter
    chain.add SidekiqUniqueJobs::Middleware::Server
    chain.add Sidekiq::Status::ServerMiddleware, expiration: 30.minutes
  end
  config.client_middleware do |chain|
    chain.add SidekiqUniqueJobs::Middleware::Client
    chain.add Sidekiq::Status::ClientMiddleware, expiration: 30.minutes
  end

  config.failures_max_count = 100

  SidekiqUniqueJobs::Server.configure(config)

  if (database_url = ENV["DATABASE_URL"])
    pool_size = 4
    ENV["DATABASE_URL"] = "#{database_url}?pool=#{pool_size}"
    ActiveRecord::Base.establish_connection
  end

end

SidekiqUniqueJobs.configure do |config|
  config.enabled = true
  config.debug_lua = true
  config.lock_info = true
  config.max_history = 1000
  config.reaper = :ruby
  config.reaper_count = 1000
  config.reaper_interval = 10
  config.reaper_timeout = 5
  config.lock_timeout = 1.minute
  config.lock_ttl = 30.minutes
end
Sidekiq::Throttled.setup!
```

There's no backtrace in the logs, when this happens it doesn't even run the first line of the worker, which outputs a logging statement.

We are not able to recreate it reliably in development, what we have observed is:

- Changing the code can sometimes cause this to happen. Queue up a number of jobs, stop the processes, change the code, start the processes again. Sometimes sidekiq will drop most of the jobs that were queued, sometimes it won't. This isn't what happened in production though as we didn't do any new code deployments.
- We had an issue with the Postgresql database that caused the jobs to run very slowly as they were waiting on locks being released in the database. There may have been a correlation between the jobs getting really slow and them not performing any actions when executed.

**Comment by mperham** [comment]
You're using so many plugins with no way to reproduce that I'm not sure I can help here. I noticed sidekiq-throttled, which replaces Sidekiq's fetcher. How do you know it's not at fault?

**Comment by egeek** [comment]
I don't know if Sidekiq throttled isn't at fault @mperham , I will take it out and re-test. I'll also add some additional logging to `job_logger` in a branch to see what exactly is going on when it yields:

```
      start = ::Process.clock_gettime(::Process::CLOCK_MONOTONIC)
      @logger.info("start")

      yield

      Sidekiq::Context.add(:elapsed, elapsed(start))
      @logger.info("done")
```

as it seems like we're at least getting the log statements before and after the `yield`, but nothing is happening in between.

**Comment by mperham** [comment]
Try 658?

**Comment by mbar0** [comment]
@egeek I have seen this behavior before associated with the sidekiq-unique-jobs gem when lock_ttl is exceeded for until_and_while_executing jobs while they are still waiting on the queue -- see: https://github.com/mhenrixon/sidekiq-unique-jobs/discussions/731

I can't speak to whether this is what is impacting you, but that might help with troubleshooting. In our case we removed the lock_ttl and rely on the sidekiq-unique-jobs reaper to clean up orphaned locks.

**Comment by courtsimas** [comment]
@egeek same thing happens for us. We'll queue up 2000 jobs in about 5 seconds, and about a quarter of them will process. We also log the first line of the jobs. We're on the latest sidekiq 6; haven't tried 7 yet. Also on ruby 3.1.2 and rails 7.0.4. It's bizarre. It's been happening for a few months, but I can't quite figure out when we exactly started seeing it. No errors, no failures, and our sidekiq job count doesn't go up by the amount of jobs it should.... it's just like some get dropped silently.

**Comment by egeek** [comment]
> @egeek I have seen this behavior before associated with the sidekiq-unique-jobs gem when lock_ttl is exceeded for until_and_while_executing jobs while they are still waiting on the queue -- see: [mhenrixon/sidekiq-unique-jobs#731](https://github.com/mhenrixon/sidekiq-unique-jobs/discussions/731)

This could be it, we have `lock_ttl` configured to be 30 minutes, which would align with the behaviour we are experiencing.

Unfortunately we don't get a good solid test run until the beginning of each month, but I'll write a test to simulate the behaviour by scheduling a large number of jobs that each sleep between 10 and 15 seconds. If that successfully recreates the behaviour then I'll try the following tests and report back:

- Upgrade to sidekiq 6.5.8
- Change the `lock_ttl` setting in unique jobs to 60 minutes (see if the dropped jobs start happening after 60 minutes), then try removing the `lock_ttl` setting altogether

**Comment by egeek** [comment]
Ran the tests. Upgrading to 6.5.8 didn't change the behaviour.

Changing `lock_ttl` did. I reduced it to 2 minutes, and jobs started being dropped from the queue after 2 minutes.

Tested again with `perform_bulk`, `perform_async` and `perform_in`, it's the same behaviour with all of them. Once the job has been in the queue for more than the `lock_ttl` limit, it's dropped.

@mperham Definitely does mean the issue is in SidekiqUniqueJobs and not Sidekiq itself (awesome work on Sidekiq BTW)
@mbar0 thanks for the steer!
@courtsimas if you are using SidekiqUniqueJobs as well, then it might be worth checking if you have `lock_ttl` configured
